cyraft/node.py

```diff
@@ -395,11 +397,14 @@ def _append_entries_processing(
         This method is called when the append entries request passes all checks and can be appended to the log.
         """
         assert len(request.log_entry) == 1  # in our implementation only a single entry is sent at a time
+        new_index = request.prev_log_index + 1
+        _logger.info(c["append_entries"] + "Node ID: %d -- new_index: %d" + c["end_color"], self._node.id, new_index)
 
+        #a leader never overwrites or deletes entries in its log; it only appends new entries
```

Comment:
[The paper](https://raft.github.io/raft.pdf) states that the leader can only add a new entity, but not rewrite or delete:
![Figure 3](https://github.com/user-attachments/assets/448979c8-0034-4095-9f24-9d7264826353)

Plus it is written the same paper that entity can only be rewritten if the index is the same and the term is different, but the term only changes after the elections, when a new leader comes.
![image](https://github.com/user-attachments/assets/67a3dd47-4d5b-45bc-a04c-293251e045e3)

Reply:
> Plus it is written the same paper that entity can only be rewritten if the index is the same and the term is different, but the term only changes after the elections, when a new leader comes.

But then why mention this at all in the implementation?

![image](https://github.com/user-attachments/assets/830eabcb-944d-438d-b90e-0be1f636e17c)

I'm pretty sure that this rule is necessary in some bordercase where a node (or a group of nodes) becomes disconnected from the larger network, they proceed to choose a leader, updates its log to something and then reconnects to the larger network (something like that). In that case it is probably necessary that the leader is able to overwrite the log of its followers.

Reply:
Yes. But the implementation of this function was incorrect. Now, every time entities are added to the log, it starts a cycle that checks each element in the log. This is not efficient, because if we have 20 nodes with 10,000 logs and we add another entity, then to put it mildly, it is not very fast, especially for RTOS

Reply:
inefficient != incorrect 

Log comparison could indeed be done more efficiently, so if you want to optimize it, go ahead.